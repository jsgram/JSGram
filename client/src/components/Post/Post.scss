@import '../../styles/variables';

.dots {
  padding-bottom: 0;
  font-weight: bold;
  font-size: 1.5em;
  line-height: 3px;
  cursor: default;
}

.navig {
  position: absolute;
  left: -50vw;
  width: 20vw;
  border: 0.5px solid $color-grey;
  border-radius: 0.3em;

  .menu-link {
    color: $color-black;
    text-decoration: none;
    border-bottom: 2px solid $color-grey;

    &:hover {
      color: $color-black;
      font-weight: bold;
      text-decoration: none;
    }
  }
}


.show-menu {
  left: 0;
}






