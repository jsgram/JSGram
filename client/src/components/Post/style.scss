@import '../../styles/variables';

  .post-photo {
    width: 100%;
    position: relative;

    &:hover {
      .post-icon {
        display: block;
      }
    }
  }

  .show-photo-like {
    cursor: pointer;
    opacity: 1;
    transition-duration: 0.5s;

    &:hover {
      opacity: 0.9;
    }
  }

  .post-icon {
    transition-duration: 0.5s;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    display: none;
    font-size: 1.2rem;
    color: $color-alabaster;
    font-weight: 600;
  }

  .like {
    color: $color-like;
  }

  .text-description {
    word-wrap: break-word;
  }

  .comments {
    height: 350px;
    @media screen and (max-width: $break-md) {
      height: 130px;
    }
  }

  .add-comment {
    height: 30px;
    max-height: 80px;
    outline: 0;
    resize: none;
  }

  .button-comment {
    background-color: $color-alabaster;
    color: $color-danube;

    &:disabled {
      opacity: 0.3;
      pointer-events: none;
    }
  }




