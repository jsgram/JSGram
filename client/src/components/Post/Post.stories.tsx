import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../../App';
import Post from './index';

const loggedUser = {
    isAdmin: false,
    loggedId: '1a',
    loggedPhotoPath: 'https://jsgram-profile-images.s3.eu-west-3.amazonaws.com/1569138592191',
    loggedUsername: 'Maks',
};

const user = {
    description: 'User description',
    email: 'maks@gmail.com',
    followers: [1, 2, 3],
    following: [1, 2, 3],
    fullName: 'Maks',
    photo: 'https://jsgram-profile-images.s3.eu-west-3.amazonaws.com/1569138592191',
    posts: 1,
    privacy: {
        isActivityStatus: true,
        isPrivateAccount: false,
        isStorySharing: true,
    },
    subscriptions: {
        isNewsEmail: true,
        isProductEmail: true,
        isReminderEmail: true,
        isResearchEmail: true,
        isTextMessage: false,
    },
    username: 'Maks',
    _id: '1a',
};

const userPosts = {
    commentsLoaded: false,
    commentsLoading: false,
    commentsPage: 1,
    countOfLikes: 0,
    likeExist: false,
    loaded: false,
    loading: false,
    page: 1,
    posts: [
        {
            author: '1a',
            authorsOfLike: ['1f', '2f', '3f'],
            comments: ['1c', '2c', '3c'],
            description: '',
            imgPath: 'https://jsgram-post-images1.s3.eu-west-3.amazonaws.com/1567816427997',
            tags: [],
            _id: '1p',
        },
    ],
    selectedPost: {
        _id: '1p',
        author: '1a',
        description: 'Post description',
        imgPath: 'https://jsgram-post-images1.s3.eu-west-3.amazonaws.com/1567816427997',
    },
};

export const postState = {
    username: 'Maks',
    loggedUser,
    user,
    userPosts,
};

export const postActions = {
    getPostsAsync: action('getPostsAsync'),
    getMorePostsAsync: action('getMorePostsAsync'),
    deletePhoto: action('deletePhoto'),
    editPost: action('editPost'),
    showPost: action('showPost'),
    getUser: action('getUser'),
    resetPosts: action('resetPosts'),
    addNextPosts: action('addNextPosts'),
    changeEditStatus: action('changeEditStatus'),
    editDescriptionForPost: action('editDescriptionForPost'),
    newDescriptionForPost: action('newDescriptionForPost'),
};

storiesOf('Post', module)
    .addDecorator((story: any): any => <Provider store={store}>{story()}</Provider>)
    .addDecorator((story: any): any => <MemoryRouter>{story()}</MemoryRouter>)
    .add('post', () => <Post {...{...postState, ...postActions}}/>);
