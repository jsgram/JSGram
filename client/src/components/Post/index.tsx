import React from 'react';
import '../../styles/style.scss';
import { Waypoint } from 'react-waypoint';
import TextareaAutosize from 'react-textarea-autosize';
import { IUserData } from '../Profile';
import { IPost } from '../../store/post/reducers';
import { Modal, ModalHeader, Spinner, Input, FormGroup, Button, InputGroup, InputGroupAddon } from 'reactstrap';
import './style.scss';
import MenuPost from '../MenuPost';
import noAvatar from '../../assets/noAvatar.svg';

interface IBody {
    userId: string;
    postId: string;
}

interface IProps {
    userPosts: any;
    user: IUserData;
    editDescriptionForPost: any;
    getPostsAsync: (username: string) => void;
    getMorePostsAsync: (username: string, page: number) => void;
    deletePhoto: () => void;
    addLike: (body: IBody) => void;
    setCountOfLikes: (countOfLikes: number) => void;
    deleteLike: (body: IBody) => void;
    countOfLikes: number;
    editPost: (description: string, id: string) => void;
    showPost: (post: any) => void;
    likeExist: boolean;
    checkUserLikeExist: (doesExist: boolean) => void;
    username: string;
    getUser: (username: string) => void;
    resetPosts: () => void;
    addNextPosts: (pageNumber: number) => void;
    loggedUsername: string;
}

interface IModalState {
    modal: boolean;
    editModal: boolean;
}

export default class Post extends React.Component<IProps> {

    public state: IModalState = {
        modal: false,
        editModal: false,
    };

    public toggle = (post: any): any => {
        this.setState({
            modal: !this.state.modal,
        });
        this.props.showPost(post);
    }

    public toggleEdit = (post: any): any => {
        this.setState({
            editModal: !this.state.editModal,
            modal: !this.state.modal,
        });
        this.props.showPost(post);
    }

    public onEditPost = (): void => {
        this.props.editPost(this.props.userPosts.selectedPost.description, this.props.userPosts.selectedPost._id);
        this.toggleEdit(this.props.userPosts.selectedPost);
    }

    public onDescriptionChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
        this.props.editDescriptionForPost(event.target.value);
    }

    public getMorePosts = (): void => {
        if (!this.props.userPosts.loaded) {
            this.props.addNextPosts(this.props.userPosts.page + 1);
            this.props.getMorePostsAsync(this.props.user.username, this.props.userPosts.page);
        }
    }

    public componentDidMount(): void {
        this.setState({page: 1});
        this.props.getPostsAsync(this.props.user.username);
    }

    public onDeleteLike = (): void => {
        const {
            user: {_id: userId},
            userPosts: {selectedPost: {_id: postId}},
        }: any = this.props;
        const body = {userId, postId};
        // TODO CAN NOT CHANGE PROPS DIRECTLY
        const index = this.props.userPosts.selectedPost.authorsOfLike.indexOf(body.userId);
        this.props.userPosts.selectedPost.authorsOfLike.splice(index, 1);
        this.props.deleteLike(body);
    }

    public onAddLike = (): void => {
        const {
            user: {_id: userId},
            userPosts: {selectedPost: {_id: postId}},
        }: any = this.props;
        const body = {userId, postId};
        // TODO CAN NOT CHANGE PROPS DIRECTLY
        this.props.userPosts.selectedPost.authorsOfLike.push(this.props.user._id);
        this.props.addLike(body);
    }

    // TODO refactoring in next sprint
    public setLikesCount = (): boolean | void => {
        if (!!this.props.userPosts.selectedPost.authorsOfLike) {
            this.props.setCountOfLikes(this.props.userPosts.selectedPost.authorsOfLike.length);

            const arr = this.props.userPosts.selectedPost.authorsOfLike.filter((userId: string) =>
                this.props.user._id === userId,
            );

            if (arr.length) {
                this.props.checkUserLikeExist(true);
                return true;
            }
            this.props.checkUserLikeExist(false);
            return false;
        }
    }

    public render(): JSX.Element {
        const { userPosts, user, likeExist, countOfLikes }: any = this.props;
        const { selectedPost: { description: desc } }: any = userPosts;

        const HASH_REGEXP = /[#][a-z]+/;
        const MENTION_REGEXP = /[@][a-z]+/;
        const LINK_REGEXP = /(?:(?:https?|ftp):\/\/|www\.)[^\s\/$.?#].[^\s]*/;

        const hashtagRegex = new RegExp(`(${HASH_REGEXP.source}|${MENTION_REGEXP.source}|${LINK_REGEXP.source})`, 'ig');
        const formatDescription = desc && desc.split(hashtagRegex).map((token: string) => {
            switch (true) {
                case !!token.match(HASH_REGEXP):
                    return (<a href={`/profile/${token.slice(1)}`}>{token}</a>);
                case !!token.match(MENTION_REGEXP):
                    return (<a href={`/profile/${token.slice(1)}`}>{token}</a>);
                case !!token.match(LINK_REGEXP):
                    return (<a href={token}>{token}</a>);
                default:
                    return token;
            }
        });

        const likeButton = this.setLikesCount() && likeExist ?
            (<i className='fa fa-heart fa-lg pr-1 like' onClick={this.onDeleteLike}/>) :
            (<i className='fa fa-heart-o fa-lg pr-1' onClick={this.onAddLike}/>);

        return (
            <div className='container justify-content-center'>
                <div className='row mt-5 profile-post'>
                    {
                        userPosts.posts.map((post: IPost) => (
                                <div key={post._id} className='col-sm-4 text-center pt-4 post-photo'>
                                    <img
                                        src={post.imgPath}
                                        width={293}
                                        height={293}
                                        alt=''
                                        onClick={(): void => this.toggle(post)}
                                        className='img-fluid one-profile-photo'
                                    />
                                </div>
                            ),
                        )
                    }
                    <Waypoint
                        scrollableAncestor={window}
                        onEnter={(): void => {
                            this.getMorePosts();
                        }}
                    />
                </div>
                <div className='w-100 d-flex align-items-center justify-content-center'>
                    { userPosts.loading && <Spinner className='mt-3' color='dark'/>}
                </div>
                <Modal className='profile-post modal-lg modal-dialog-centered px-3 py-3'
                       isOpen={this.state.modal}
                       toggle={(): void => this.toggle(userPosts.selectedPost)}>
                    <div className='modal-body p-0'>
                        <div className='row m-0'>
                            <ModalHeader
                                className='d-lg-none w-100 display-1'
                                toggle={(): void => this.toggle(userPosts.selectedPost)}
                            >
                                <div className='d-flex'>
                                    <MenuPost
                                        post={userPosts.selectedPost}
                                        toggleEdit={this.toggleEdit}
                                        toggleModal={this.toggle}
                                    />
                                    <img
                                        src={user.photo || noAvatar}
                                        alt='avatar'
                                        width={32}
                                        height={32}
                                        className='img-fluid rounded-circle mt-2 ml-4'
                                    />
                                    <span className='mt-2 ml-2'>{user.username}</span>
                                </div>
                            </ModalHeader>

                            <div className='col-12 col-lg-8 px-0'>
                                <img
                                    src={userPosts.selectedPost.imgPath}
                                    className='w-100 img-fluid'
                                    alt='post'
                                />
                            </div>

                            <div className='col-12 col-lg-4 px-0 d-flex flex-column position-relative'>
                                <div className='flex-grow-0 p-3 text-description'>
                                    <div className='d-lg-block d-none flex-row'>
                                        <img
                                            src={user.photo || noAvatar}
                                            alt='avatar'
                                            width={32}
                                            height={32}
                                            className='img-fluid rounded-circle mt-2 mr-2 mb-2'
                                        />
                                        <span className='mt-2 font-weight-bolder'>{user.username}</span>
                                        <div className='d-lg-block d-none float-right'>
                                            <MenuPost
                                                post={userPosts.selectedPost}
                                                toggleEdit={this.toggleEdit}
                                                toggleModal={this.toggle}
                                            />
                                        </div>
                                    </div>
                                    <p className='d-lg-none'>
                                        {likeButton}
                                        <span>{countOfLikes} likes</span>
                                    </p>
                                    <p>{formatDescription}</p>
                                </div>

                                <div className='flex-grow-1 comments px-3 text-description'>
                                    {/* HERE WILL BE COMMENTS */}
                                </div>

                                <div className='flex-grow-0'>
                                    <div className='d-none d-lg-block p-3 mb-3 border-top border-bottom'>
                                        {likeButton}
                                        <span>{countOfLikes} likes</span>
                                    </div>
                                    <InputGroup>
                                        <TextareaAutosize
                                            className='add-comment flex-grow-1 border-0 p-2'
                                            placeholder='Write your comment...'
                                            autoComplete='off'
                                            minRows={1}
                                            maxRows={4}
                                        />
                                        <InputGroupAddon addonType='append' className='flex-grow-0'>
                                            <Button
                                                className='btn-block button-comment border-0'
                                                type='submit'
                                                disabled
                                            >
                                                Add
                                            </Button>
                                        </InputGroupAddon>
                                    </InputGroup>
                                </div>
                            </div>
                        </div>
                    </div>
                </Modal>

                <Modal
                    isOpen={this.state.editModal}
                    toggle={(): void => this.toggleEdit(userPosts.selectedPost)}>
                    <ModalHeader
                        toggle={(): void => this.toggleEdit(userPosts.selectedPost)}>
                        Edit Post
                    </ModalHeader>
                    <FormGroup className='text-center m-3 post-photo'>
                        <img
                            src={userPosts.selectedPost.imgPath}
                            width={293}
                            height={293}
                            id={userPosts.selectedPost._id}
                            alt='post'
                            className='img-fluid w-100'
                        />
                    </FormGroup>
                    <FormGroup className='m-3 change-description'>
                        <Input
                            type='textarea'
                            name='description'
                            placeholder='Write a caption...'
                            spellCheck={false}
                            value={userPosts.selectedPost.description}
                            onChange={this.onDescriptionChange}
                        />
                        <Button
                            color='danger'
                            className='mt-2'
                            block
                            onClick={this.onEditPost}>
                            Update Post
                        </Button>
                    </FormGroup>
                </Modal>
            </div>
        );
    }
}
