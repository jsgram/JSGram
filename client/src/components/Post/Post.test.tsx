import React from 'react';
import Post from './index';
import { shallow } from 'enzyme';

describe('Post component', () => {
    let renderer: any;
    const props = {
        userPosts: {
            selectedPost: {
                description: 'some value',
            },
            posts: [],
        },
        user: {
            posts: 1,
            followers: [],
            following: [],
            description: 'some value',
            fullName: 'some value',
            username: 'some value',
            photo: 'some value',
            email: 'some value',
            _id: 'some value',
        },
        editDescriptionForPost: 'some value',
        newDescriptionForPost: 'some value',
        username: 'some value',
        getPostsAsync: jest.fn(() => 'some value'),
        getMorePostsAsync: jest.fn(() => 'some value'),
        deletePhoto: jest.fn(() => 'some value'),
        editPost: jest.fn(() => 'some value'),
        showPost: jest.fn(() => 'some value'),
        getUser: jest.fn(() => 'some value'),
        resetPosts: jest.fn(() => 'some value'),
        addNextPosts: jest.fn(() => 'some value'),
        loggedUser: {
            loggedUsername: 'some value',
            loggedId: 'some value',
            loggedPhotoPath: 'some value',
            isAdmin: true,
        },
        changeEditStatus: jest.fn(() => 'some value'),
    };

    beforeEach(() => {
        renderer = shallow(<Post {...props}/>);
    });

    test('toggle - success', () => {
        renderer.instance().toggle();
        expect(props.showPost).toHaveReturnedWith('some value');
    });

    test('toggle edit - success', () => {
        renderer.instance().toggleEdit({});
        expect(props.showPost).toHaveReturnedWith('some value');
        expect(props.changeEditStatus).toHaveReturnedWith('some value');
    });

    test('edit post - success', () => {
        renderer.instance().onEditPost();
        expect(props.editPost).toHaveReturnedWith('some value');
    });

    test('get more posts - success', () => {
        renderer.instance().getMorePosts();
        expect(props.addNextPosts).toHaveReturnedWith('some value');
        expect(props.getMorePostsAsync).toHaveReturnedWith('some value');
    });

    test('componentDidMount - success', () => {
        renderer.instance().componentDidMount();
        expect(props.getPostsAsync).toHaveReturnedWith('some value');
    });

    test('render - success', () => {
        expect(renderer).toMatchSnapshot();
    });
});
