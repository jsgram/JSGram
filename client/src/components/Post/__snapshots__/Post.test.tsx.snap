// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post component render - success 1`] = `ShallowWrapper {}`;
